public/hooks/use_chat_state.tsx

```diff
@@ -76,6 +110,20 @@ const chatStateReducer: React.Reducer<ChatState, ChatStateAction> = (state, acti
         draft.llmResponding = true;
         draft.llmError = undefined;
         break;
+      /**
+       * As in Olly, regenerate and send_message API will only response with the latest one interaction and messages to improve performance.
+       * So we need to use patch to hanlde with the response.
+       */
+      case 'patch':
```

Comment:
Seems the `patch` action always follow `receive` action. Shall we just merge `patch` action to `receive` action?

Reply:
> the patch action always follow receive action

For now, yes because for send_message and regenerate action, we append dummy input message and when we do patch, we need to call `receive` to remove the dummy message. But there are cases that we only call receive action without patch action.

> Shall we just merge patch action to receive action

I'd say no because patch action can only do add/update, not able to "remove" a useless message/interaction. Only receive action can do that.

Reply:
You're right. I missed the case in `loadChat`. The `loadChat` will call receive action without patch.
My original thoughts was to refactor the receive action to accept patch messages and interactions like below code:
```ts
    dispatch({
        type: 'receive',
        payload: {
          messages: response.messages,
          interactions: response.interactions,
          patch: {messages, interactions}
        },
      });
```
We can add an optional patch parameter, and complete patch inside receive action.

Reply:
I am thinking if that's too much for a single action to do `set` as well as `patch`.